sycl/source/detail/device_binary_image.cpp

```diff
@@ -235,6 +250,420 @@ DynRTDeviceBinaryImage::~DynRTDeviceBinaryImage() {
   Bin = nullptr;
 }
 
+// Exclusive property merge logic. It assumes there are no cases where
+// properties have different values and throws otherwise.
+template <typename RangeGetterT>
+static std::vector<sycl_device_binary_property>
+NaiveMergeBinaryProperties(const std::vector<const RTDeviceBinaryImage *> &Imgs,
+                           const RangeGetterT &RangeGetter) {
+  size_t PropertiesCount = 0;
+  for (const RTDeviceBinaryImage *Img : Imgs)
+    PropertiesCount += RangeGetter(*Img).size();
+
+  std::vector<sycl_device_binary_property> Props;
+  Props.reserve(PropertiesCount);
+  for (const RTDeviceBinaryImage *Img : Imgs) {
+    const RTDeviceBinaryImage::PropertyRange &Range = RangeGetter(*Img);
+    Props.insert(Props.end(), Range.begin(), Range.end());
+  }
+
+  return Props;
+}
+
+// Exclusive property merge logic. If IgnoreDuplicates is false it assumes there
+// are no cases where properties have different values and throws otherwise.
+template <typename RangeGetterT>
+static std::map<std::string_view, const sycl_device_binary_property>
+ExclusiveMergeBinaryProperties(
+    const std::vector<const RTDeviceBinaryImage *> &Imgs,
+    const RangeGetterT &RangeGetter, bool IgnoreDuplicates = false) {
+  std::map<std::string_view, const sycl_device_binary_property> MergeMap;
+  for (const RTDeviceBinaryImage *Img : Imgs) {
+    const RTDeviceBinaryImage::PropertyRange &Range = RangeGetter(*Img);
+    for (const sycl_device_binary_property Prop : Range) {
+      const auto [It, Inserted] =
+          MergeMap.try_emplace(std::string_view{Prop->Name}, Prop);
+      if (IgnoreDuplicates || Inserted)
+        continue;
+      // If we didn't insert a new entry, check that the old entry had the
+      // exact same value.
+      const sycl_device_binary_property OtherProp = It->second;
+      if (OtherProp->Type != Prop->Type ||
+          OtherProp->ValSize != Prop->ValSize ||
+          (Prop->Type == SYCL_PROPERTY_TYPE_BYTE_ARRAY &&
+           std::memcmp(OtherProp->ValAddr, Prop->ValAddr, Prop->ValSize) != 0))
+        throw sycl::exception(make_error_code(errc::invalid),
+                              "Unable to merge incompatible images.");
+    }
+  }
+  return MergeMap;
+}
+
+// Device requirements needs the ability to produce new properties. The
+// information for these are kept in this struct.
+struct MergedDeviceRequirements {
+  std::map<std::string_view, const sycl_device_binary_property> MergeMap;
+  std::set<uint32_t> Aspects;
+  std::set<std::string_view> JointMatrix;
+  std::set<std::string_view> JointMatrixMad;
```

Comment:
Do we need these to be ordered containers?

Reply:
Unordered should be fine for these.

Reply:
What about the merge map?